[JENKINS-33971] ReverseBuildTrigger.upstream2Trigger cache fix

jglick opened the pull request:
[JENKINS-33971](https://issues.jenkins-ci.org/browse/JENKINS-33971)

@reviewbybees esp. @fbelzunc

jglick committed:
[FIXED JENKINS-33971] ReverseBuildTrigger.upstream2Trigger was not being reliably calculated during Jenkins startup.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

amuniz commented on core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java:
```diff
@@ -221,9 +199,53 @@ public FormValidation doCheckUpstreamProjects(@AncestorInPath Job project, @Quer
     @Extension public static final class RunListenerImpl extends RunListener<Run> {
+        
+        static RunListenerImpl get() {
+            return ExtensionList.lookup(RunListener.class).get(RunListenerImpl.class);
```
Why not `ExtensionList.lookup(RunListenerImpl.class)`?

jglick replied:
Not sure if that even works—the extension point is `RunListener`. Either way you would need to have the call to `get` or similar anyway.

amuniz commented:
:bee:

oleg-nenashev commented on core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java:
```diff
@@ -221,9 +199,53 @@ public FormValidation doCheckUpstreamProjects(@AncestorInPath Job project, @Quer
+
+        private Map<Job,Collection<ReverseBuildTrigger>> calculateCache() {
+            Map<Job,Collection<ReverseBuildTrigger>> result = new WeakHashMap<>();
+            SecurityContext orig = ACL.impersonate(ACL.SYSTEM);
```
Are you sure that <code>finally</code> handler is placed correctly? It's inside for loop :ant:

oleg-nenashev commented:
Maybe the logic impersonation is correct, but I don't understand it in such case.
:bug: , because the second <code>for (Job<?, ?> downstream : Jenkins.getInstance().getAllItems(Job.class)) </code> iteration will be executed with the non-admin context

jglick committed:
@oleg-nenashev noticed that a finally block was misplaced.

jglick committed:
Found that ItemListenerImpl was failing to call ACL.impersonate on ACL.SYSTEM.
Also switching to the closure-based method which is not susceptible to bad nesting mistakes.

oleg-nenashev commented:
:bee:
